ambry-router/src/test/java/com.github.ambry.router/GetOperationTest.java

```diff
@@ -0,0 +1,474 @@
+import com.github.ambry.network.NetworkClientErrorCode;
+import com.github.ambry.network.RequestInfo;
+import com.github.ambry.network.ResponseInfo;
+import com.github.ambry.protocol.RequestOrResponse;
+import com.github.ambry.utils.MockTime;
+import com.github.ambry.utils.Utils;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicReference;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+
+public class GetOperationTest {
+  private static final int MAX_PORTS_PLAIN_TEXT = 3;
+  private static final int MAX_PORTS_SSL = 3;
+  private static final int CHECKOUT_TIMEOUT_MS = 1000;
+
+  private int requestParallelism = 2;
+  private int successTarget = 1;
+  private final RouterConfig routerConfig;
+  private final MockClusterMap mockClusterMap;
+  private final MockServerLayout mockServerLayout;
+  private final int replicasCount;
+  private final AtomicReference<MockSelectorState> mockSelectorState = new AtomicReference<>();
+  private final ResponseHandler responseHandler;
+  private final MockNetworkClientFactory networkClientFactory;
+  private final NetworkClient networkClient;
+  private final MockTime time;
+  private final Map<Integer, GetOperation> correlationIdToGetOperation = new HashMap<>();
+  private final NonBlockingRouter router;
+  private final int BLOB_SIZE = 100;
+  private final int BLOB_USER_METADATA_SIZE = 10;
+  private final Random random = new Random();
+  private final String blobIdStr;
+  private final BlobProperties blobProperties;
+  private final byte[] userMetadata;
+  private final byte[] putContent;
+  private final GetTestRequestRegistrationCallbackImpl requestRegistrationCallback =
+      new GetTestRequestRegistrationCallbackImpl();
+  private final FutureResult<BlobInfo> operationFuture = new FutureResult<>();
+  private final AtomicReference<BlobInfo> operationResult = new AtomicReference<>(null);
+  private final AtomicReference<Exception> operationException = new AtomicReference<>(null);
+  private final Callback<BlobInfo> operationCallback = new Callback<BlobInfo>() {
+    @Override
+    public void onCompletion(BlobInfo result, Exception exception) {
+      operationResult.set(result);
+      operationException.set(exception);
+    }
+  };
+
+  private class GetTestRequestRegistrationCallbackImpl implements RequestRegistrationCallback<GetOperation> {
+    private List<RequestInfo> requestListToFill;
+
+    @Override
+    public void registerRequestToSend(GetOperation getOperation, RequestInfo requestInfo) {
+      requestListToFill.add(requestInfo);
+      correlationIdToGetOperation.put(((RequestOrResponse) requestInfo.getRequest()).getCorrelationId(), getOperation);
+    }
+  }
+
+  ;
+
+  public GetOperationTest()
+      throws Exception {
```

Comment:
Again, I don't think that is necessary. The caller does not care. As for readability, exceptions only get thrown if the tests fail, and I am not sure if there is value in knowing what exceptions get thrown if the tests fail.